<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read exposure or outcome data — format_data • TwoSampleMR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Read exposure or outcome data — format_data" />
<meta property="og:description" content="Reads in exposure data. Checks and organises columns for use with MR or enrichment tests. Infers p-values when possible from beta and se. If it is the exposure then looks up SNPs in biomaRt to get basic info." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TwoSampleMR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/index.html">Guide</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/mrcieu/TwoSampleMR">Source</a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read exposure or outcome data</h1>
    
    <div class="hidden name"><code>format_data.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Reads in exposure data. Checks and organises columns for use with MR or enrichment tests. Infers p-values when possible from beta and se. If it is the exposure then looks up SNPs in biomaRt to get basic info.</p>
    </div>

    <pre class="usage"><span class='fu'>format_data</span>(
  <span class='no'>dat</span>,
  <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"exposure"</span>,
  <span class='kw'>snps</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>header</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>phenotype_col</span> <span class='kw'>=</span> <span class='st'>"Phenotype"</span>,
  <span class='kw'>snp_col</span> <span class='kw'>=</span> <span class='st'>"SNP"</span>,
  <span class='kw'>beta_col</span> <span class='kw'>=</span> <span class='st'>"beta"</span>,
  <span class='kw'>se_col</span> <span class='kw'>=</span> <span class='st'>"se"</span>,
  <span class='kw'>eaf_col</span> <span class='kw'>=</span> <span class='st'>"eaf"</span>,
  <span class='kw'>effect_allele_col</span> <span class='kw'>=</span> <span class='st'>"effect_allele"</span>,
  <span class='kw'>other_allele_col</span> <span class='kw'>=</span> <span class='st'>"other_allele"</span>,
  <span class='kw'>pval_col</span> <span class='kw'>=</span> <span class='st'>"pval"</span>,
  <span class='kw'>units_col</span> <span class='kw'>=</span> <span class='st'>"units"</span>,
  <span class='kw'>ncase_col</span> <span class='kw'>=</span> <span class='st'>"ncase"</span>,
  <span class='kw'>ncontrol_col</span> <span class='kw'>=</span> <span class='st'>"ncontrol"</span>,
  <span class='kw'>samplesize_col</span> <span class='kw'>=</span> <span class='st'>"samplesize"</span>,
  <span class='kw'>gene_col</span> <span class='kw'>=</span> <span class='st'>"gene"</span>,
  <span class='kw'>id_col</span> <span class='kw'>=</span> <span class='st'>"id"</span>,
  <span class='kw'>min_pval</span> <span class='kw'>=</span> <span class='fl'>1e-200</span>,
  <span class='kw'>z_col</span> <span class='kw'>=</span> <span class='st'>"z"</span>,
  <span class='kw'>info_col</span> <span class='kw'>=</span> <span class='st'>"info"</span>,
  <span class='kw'>chr_col</span> <span class='kw'>=</span> <span class='st'>"chr"</span>,
  <span class='kw'>pos_col</span> <span class='kw'>=</span> <span class='st'>"pos"</span>,
  <span class='kw'>log_pval</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dat</th>
      <td><p>Data frame. Must have header with at least SNP column present.</p></td>
    </tr>
    <tr>
      <th>type="exposure".</th>
      <td><p>Is this the exposure or the outcome data that is being read in?</p></td>
    </tr>
    <tr>
      <th>snps=NULL</th>
      <td><p>SNPs to extract. If NULL then doesn't extract any and keeps all.</p></td>
    </tr>
    <tr>
      <th>phenotype_col="Phenotype"</th>
      <td><p>Optional column name for the column with phenotype name corresponding the the SNP. If not present then will be created with the value "Outcome"</p></td>
    </tr>
    <tr>
      <th>snp_col="SNP"</th>
      <td><p>Required name of column with SNP rs IDs</p></td>
    </tr>
    <tr>
      <th>beta_col="beta"</th>
      <td><p>Required for MR. Name of column with effect sizes</p></td>
    </tr>
    <tr>
      <th>se_col="se"</th>
      <td><p>Required for MR. Name of column with standard errors</p></td>
    </tr>
    <tr>
      <th>eaf_col="eaf"</th>
      <td><p>Required for MR. Name of column with effect allele frequency</p></td>
    </tr>
    <tr>
      <th>effect_allele_col="effect_allele"</th>
      <td><p>Required for MR. Name of column with effect allele. Must contain only the characters "A", "C", "T" or "G"</p></td>
    </tr>
    <tr>
      <th>other_allele_col="other_allele"</th>
      <td><p>Required for MR. Name of column with non effect allele. Must contain only the characters "A", "C", "T" or "G"</p></td>
    </tr>
    <tr>
      <th>pval_col="pval"</th>
      <td><p>Required for enrichment tests. Name of column with p-value.</p></td>
    </tr>
    <tr>
      <th>units_col="units"</th>
      <td><p>Optional column name for units.</p></td>
    </tr>
    <tr>
      <th>ncase_col="ncase"</th>
      <td><p>Optional column name for number of cases.</p></td>
    </tr>
    <tr>
      <th>ncontrol_col="ncontrol"</th>
      <td><p>Optional column name for number of controls.</p></td>
    </tr>
    <tr>
      <th>samplesize_col="samplesize"</th>
      <td><p>Optional column name for sample size.</p></td>
    </tr>
    <tr>
      <th>gene_col="gene"</th>
      <td><p>Optional column name for gene name.</p></td>
    </tr>
    <tr>
      <th>min_pval=1e-200</th>
      <td><p>Minimum allowed pval</p></td>
    </tr>
    <tr>
      <th>log_pval=FALSE</th>
      <td><p>The pval is -log10(P)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>data frame</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gibran Hemani, Philip Haycock, Jie Zheng, Tom Gaunt, Ben Elsworth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


